rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,456 @@
+---
+title: "Router Custom Module System v1"
+author: Dustin Deus
+date: 2024-08-11
+status: Draft
+---
+
+# Router Custom Module System v1
+
+- **Author:** Dustin Deus
+- **Date:** 2024-08-11
+- **Status:** Draft
+
+## Abstract
+
+This RFC describes an overhaul of the current module system in the router. The new module system is designed to be more flexible and native to GraphQL. It allows developers to hook into the gateway lifecycle as well as outgoing and incoming requests to subgraphs.
+
+## Introduction
+
+As of today, customers can extend the router with custom modules. These modules can be used to change the behavior of the router, add custom logic, or integrate with other systems. The current module system has several limitations that we want to address with this RFC:
+
+- The current module system is not native to GraphQL. It is based on HTTP middleware and does not provide a GraphQL-specific API.
+- The current module system is inconsistent and hard to use. It does not provide a clear API for developers to intercept and modify GraphQL requests and responses.
+- The current module system does not provide a way to create or modify OpenTelemetry data, logs for different parts of the gateway lifecycle.
+- The current module system does not provide a way to interact with the parsed, normalized, and planned GraphQL operation in order to implement custom logic.
+
+Ultimately, custom modules must be self-contained, composable and testable. They should provide a clear API for developers to interact with the gateway and subgraph lifecycle and implement custom logic without having to understand the internal workings of the router or advanced Go programming concepts.
```

Comment:
Should we use the term "gateway" or "Router"?

Reply:
I like Gateway. Router is the whole product, gateway represents for me the first layer when a request hits the router. After that, it will be passed to subgraphs.

Reply:
Renamed to router